File: src/sc_neurocore/accel/julia/neurons/non_resetting_lif.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li

"""Source-faithful Kobayashi MAT(1) non-resetting neuron."""
module NonResettingLifAccel

export step!, simulate, NonResettingLIFNeuronState, valid, reset!, threshold

"""Complete MAT(1) state and documented numerical specialization."""
mutable struct NonResettingLIFNeuronState
    v::Float64
    theta::Float64
    refractory_remaining::Float64
    omega::Float64
    tau_m::Float64
    tau_theta::Float64
    alpha::Float64
    resistance::Float64
    refractory_period::Float64
    dt::Float64
end

"""Construct the enrolled MAT(1) specialization."""
function NonResettingLIFNeuronState()
    return NonResettingLIFNeuronState(0.0, 0.0, 0.0, 19.0, 5.0, 50.0, 37.0, 50.0, 2.0, 0.001)
end

"""Return the instantaneous adaptive threshold in millivolts."""
threshold(s::NonResettingLIFNeuronState)::Float64 = s.omega + s.theta

"""Return whether complete state and configuration satisfy the safety contract."""
function valid(s::NonResettingLIFNeuronState)::Bool
    return all(isfinite, (s.v, s.theta, s.refractory_remaining, s.omega, s.tau_m, s.tau_theta, s.alpha, s.resistance, s.refractory_period, s.dt)) &&
        -200.0 <= s.v <= 200.0 && 0.0 <= s.theta <= 1.0e9 && abs(s.omega) <= 1.0e9 &&
        0.0 <= s.alpha <= 1.0e9 && s.tau_m > 0.0 && s.tau_theta > 0.0 && s.resistance > 0.0 &&
        s.refractory_period >= 0.0 && s.dt > 0.0 && 0.0 <= s.refractory_remaining <= s.refractory_period
end

"""Advance one atomic source MAT(1) sample."""
function step!(s::NonResettingLIFNeuronState, current::Float64=0.0; dt::Float64=s.dt)::Int
    if !isfinite(current) || !isfinite(dt) || dt <= 0.0 || !valid(s)
        throw(DomainError((s.v, s.theta, current, dt), "invalid MAT(1) state, configuration, current, or timestep"))
    end
    next_v = s.v + dt * (-s.v + s.resistance * current) / s.tau_m
    next_theta = s.theta * exp(-dt / s.tau_theta)
    next_refractory = max(0.0, s.refractory_remaining - dt)
    if !all(isfinite, (next_v, next_theta, next_refractory)) || !(-200.0 <= next_v <= 200.0) || !(0.0 <= next_theta <= 1.0e9)
        throw(DomainError((next_v, next_theta, next_refractory), "MAT(1) candidate left the safety envelope"))
    end
    spike = next_refractory == 0.0 && next_v >= s.omega + next_theta
    if spike
        next_theta += s.alpha
        next_refractory = s.refractory_period
    end
    if !isfinite(next_theta) || next_theta > 1.0e9
        throw(DomainError(next_theta, "MAT(1) post-event threshold left the safety envelope"))
    end
    s.v = next_v
    s.theta = next_theta
    s.refractory_remaining = next_refractory
    s.dt = dt
    return spike ? 1 : 0
end

"""Restore zero-rest source state while retaining configuration."""
function reset!(s::NonResettingLIFNeuronState)::Nothing
    s.v = 0.0; s.theta = 0.0; s.refractory_remaining = 0.0
    return nothing
end

"""Simulate a configured MAT(1) current vector and return complete traces."""
function simulate(currents::AbstractVector{<:Real}; state::NonResettingLIFNeuronState=NonResettingLIFNeuronState())
    voltages = Vector{Float64}(undef, length(currents))
    thresholds = similar(voltages)
    refractory = similar(voltages)
    events = Vector{Int64}(undef, length(currents))
    for index in eachindex(currents)
        events[index] = step!(state, Float64(currents[index]); dt=state.dt)
        voltages[index] = state.v
        thresholds[index] = state.theta
        refractory[index] = state.refractory_remaining
    end
    return (; voltages, theta=thresholds, refractory, events, state)
end

"""Simulate a constant-current MAT(1) trace."""
function simulate(n_steps::Int=1000; current::Float64=0.7, dt::Float64=0.001)
    s = NonResettingLIFNeuronState(); trace = zeros(n_steps); spikes = 0
    for index in eachindex(trace)
        spikes += step!(s, current; dt=dt); trace[index] = s.v
    end
    return trace, spikes
end

end # module NonResettingLifAccel

File: src/sc_neurocore/accel/julia/neurons/non_resetting_lif.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — dedicated Julia neuron facade

from __future__ import annotations

from typing import Any

import numpy as np
import numpy.typing as npt

from ._runtime import JULIA_MAIN as _jl
from ._runtime import KERNEL_DIR as _KERNEL_DIR

_NON_RESETTING_LIF_LOADED = False
_SC_NON_RESETTING_ADAPTIVE_LIF_LOADED = False


def _ensure_non_resetting_lif_loaded() -> Any:
    """Include the source MAT(1) module into Julia Main on first use."""
    global _NON_RESETTING_LIF_LOADED
    if _jl is None:
        raise ImportError("juliacall not available; install the `julia` extra")
    if not _NON_RESETTING_LIF_LOADED:
        jl_path = _KERNEL_DIR / "non_resetting_lif.jl"
        if not jl_path.is_file():
            raise FileNotFoundError(f"non_resetting_lif.jl missing at {jl_path}")
        _jl.include(str(jl_path))
        _NON_RESETTING_LIF_LOADED = True
    return _jl.NonResettingLifAccel


def simulate_non_resetting_lif(
    currents: npt.ArrayLike,
    *,
    v: float = 0.0,
    theta: float = 0.0,
    refractory_remaining: float = 0.0,
    omega: float = 19.0,
    tau_m: float = 5.0,
    tau_theta: float = 50.0,
    alpha: float = 37.0,
    resistance: float = 50.0,
    refractory_period: float = 2.0,
    dt: float = 0.001,
) -> dict[str, object]:
    """Run a complete source MAT(1) state/event trace in Julia."""
    drive = np.ascontiguousarray(currents, dtype=np.float64)
    if drive.ndim != 1 or not np.isfinite(drive).all():
        raise ValueError("Julia MAT(1) current must be finite and one-dimensional")
    module = _ensure_non_resetting_lif_loaded()
    state = module.NonResettingLIFNeuronState(
        v,
        theta,
        refractory_remaining,
        omega,
        tau_m,
        tau_theta,
        alpha,
        resistance,
        refractory_period,
        dt,
    )
    result = module.simulate(drive, state=state)
    return {
        "voltages": np.asarray(result.voltages, dtype=np.float64),
        "theta": np.asarray(result.theta, dtype=np.float64),
        "refractory": np.asarray(result.refractory, dtype=np.float64),
        "events": np.asarray(result.events, dtype=np.int64),
        "v_final": float(result.state.v),
        "theta_final": float(result.state.theta),
        "refractory_final": float(result.state.refractory_remaining),
    }


def _ensure_sc_non_resetting_adaptive_lif_loaded() -> Any:
    """Include the retained project module into Julia Main on first use."""
    global _SC_NON_RESETTING_ADAPTIVE_LIF_LOADED
    if _jl is None:
        raise ImportError("juliacall not available; install the `julia` extra")
    if not _SC_NON_RESETTING_ADAPTIVE_LIF_LOADED:
        jl_path = _KERNEL_DIR / "sc_non_resetting_adaptive_lif.jl"
        if not jl_path.is_file():
            raise FileNotFoundError(f"sc_non_resetting_adaptive_lif.jl missing at {jl_path}")
        _jl.include(str(jl_path))
        _SC_NON_RESETTING_ADAPTIVE_LIF_LOADED = True
    return _jl.SCNonResettingAdaptiveLifAccel


def simulate_sc_non_resetting_adaptive_lif(
    currents: npt.ArrayLike,
    *,
    v: float = -65.0,
    theta: float = -50.0,
    v_rest: float = -65.0,
    theta_rest: float = -50.0,
    delta_theta: float = 5.0,
    tau_m: float = 10.0,
    tau_theta: float = 50.0,
    r_m: float = 1.0,
    dt: float = 0.1,
) -> dict[str, object]:
    """Run a complete retained-project state/event trace in Julia."""
    drive = np.ascontiguousarray(currents, dtype=np.float64)
    if drive.ndim != 1 or not np.isfinite(drive).all():
        raise ValueError("Julia SC adaptive LIF current must be finite and one-dimensional")
    module = _ensure_sc_non_resetting_adaptive_lif_loaded()
    state = module.SCNonResettingAdaptiveLIFNeuronState(
        v,
        theta,
        v_rest,
        theta_rest,
        delta_theta,
        tau_m,
        tau_theta,
        r_m,
        dt,
    )
    result = module.simulate(drive, state=state)
    return {
        "voltages": np.asarray(result.voltages, dtype=np.float64),
        "theta": np.asarray(result.theta, dtype=np.float64),
        "events": np.asarray(result.events, dtype=np.int64),
        "v_final": float(result.state.v),
        "theta_final": float(result.state.theta),
    }
